whitesourceExecuteScan: bypass UA for mta, maven, npm and yarn

stippi2 opened the pull request:
# Changes

* Pull versioning pkg usage into part of code that can be tested.
* Re-implement "npm" scan without yarn, perform previous npm-code for type "yarn".
* Write "productToken" into npm/yarn config, unless "projectToken" is already configured. (Aligns with `WhitesourceConfigurationHelper.groovy`).
* Bypass UA also for `scanType` "maven".
* Bypass UA for `scanType` "mta", by re-using the implementation of "maven" and "npm" in sequence.
* Distinguish between doing an aggregate scan with the UA, or scanning each module into a dedicated WS project.
* When scanning multiple projects (modules), poll each one, download reports for each, do report analysis for each...
* Persist scanned projects in commonPipelineEnvironment.values for other steps to use. This avoids other steps figuring out list of project names in slightly different ways.
* Introduces limited parsing of pom.xml files, in order to figure out reachable modules and sub-artifact names. This is in pkg/maven.
* yamlFile in versioning needs GroupID field, otherwise the artifact coordinates cannot be extracted with the function transforms or whatever that is called. The field is never filled, but without it being present this breaks:
```golang
    nameTmpl := `{{list .GroupID .ArtifactID | join "-" | trimAll "-"}}`
    name, version := versioning.DetermineProjectCoordinates(nameTmpl, ...)
```
* Brings coverage of the step to 72%.

- [x] Tests
- [x] Documentation

fwilhe commented on cmd/whitesourceExecuteScan.go:
```diff
@@ -366,29 +665,29 @@ func autoGenerateWhitesourceConfig(config *ScanOptions, cmd *command.Command) er
 
 	// archiveExtractionDepth=0
-	if err := cmd.RunExecutable("sed", "-ir", `s/^[#]*\s*archiveExtractionDepth=.*/archiveExtractionDepth=0/`,
+	if err := utils.RunExecutable("sed", "-ir", `s/^[#]*\s*archiveExtractionDepth=.*/archiveExtractionDepth=0/`,
```
Not introduced in this PR, but do we really need/want to depend on `sed`? this will make the piper binary less portable, even probably between linux versions with different implementations of sed, let alone macos or windows. Do we really need this? I assume the same should be possible using native go code.

stippi2 replied:
For what it is worth, I remember thinking the same thing when I came across this. Totally agree.

stippi2 commented:
Updated, merge conflicts resolved. Now only contains the new scan methods. Converted to draft-status again, since I plan to add unit-tested for those.

OliverNocon commented on cmd/whitesourceExecuteScan.go:
```diff
@@ -3,6 +3,9 @@ package cmd
 	"fmt"
+	"github.com/SAP/jenkins-library/pkg/maven"
+	"github.com/SAP/jenkins-library/pkg/npm"
+	"github.com/docker/docker/pkg/ioutils"
```
is this reference intended?

stippi2 replied:
Hm. No, I don't think so. Thanks for pointing out!

OliverNocon reviewed (commented):
I will try to test-drive with a project once out of `Draft` state ...

stippi2 commented:
> I will try to test-drive with a project once out of `Draft` state ...

The picture in my head is slowly forming, of how everything is supposed to fit together (including the subsequent PPMS check). Yesterday I realized that some parts of the step need to work differently, which has a lot to do with your idea to persist the scanned projects in the CPE. I think this needs to be done in this PR, before the step is usable at all without the UA. It doesn't currently work for our use-case, yet. Or phrased more precisely: It's too easy to break via configuration. It would currently only work when /not/ specifying "projectName" and also /not/ specifying any build descriptor excludes. But I am not sure if for that case (all modules are scanned into separate projects), if what the step downloads and checks as reports, is actually for all projects, or only for the one module that happened to be scanned last. I hope this accurately reflects my current status.

stippi2 commented:
Meanwhile I confirmed that aggregating npm modules into the same WhiteSource project never worked. The WS project then contained only results for whatever happened to be the last scanned npm module. I still make it more failsafe and will update the PR description accordingly.

@OliverNocon I'd be interested in your feedback on the functionality I've added in the maven package, "visiting" all active maven modules with a call-back function.

CCFenner commented on cmd/whitesourceExecuteScan.go:
```diff
@@ -90,89 +126,183 @@ func newWhitesourceUtils() *whitesourceUtilsBundle {
 }
 
+// whitesourceScan stores information about scanned projects
+type whitesourceScan struct {
```
What about extracting this and related functions to an own package `pkg/whitesource/scan`?

stippi2 replied:
I've started another branch "ws-package" and this portion is already done. I would like to address this in a follow up PR in order not to invalidate the reviews on this PR.

CCFenner commented on cmd/whitesourceExecuteScan.go:
```diff
@@ -403,60 +837,82 @@ func blockUntilProjectIsUpdated(config *ScanOptions, sys whitesource, currentTim
 
 // downloadReports downloads a project's risk and vulnerability reports
-func downloadReports(config *ScanOptions, utils whitesourceUtils, sys whitesource) ([]piperutils.Path, error) {
+func downloadReports(config *ScanOptions, scan *whitesourceScan, utils whitesourceUtils, sys whitesource) ([]piperutils.Path, error) {
```
What about extracting this and related functions to an own package `pkg/whitesource/report`?

CCFenner commented on cmd/whitesourceExecuteScan.go:
```diff
@@ -184,65 +314,202 @@ func resolveProjectIdentifiers(config *ScanOptions, utils whitesourceUtils, sys
 }
 
+// executeMTAScan executes a scan for the Java part with maven, and performs a scan for each NPM module.
+func executeMTAScan(config *ScanOptions, scan *whitesourceScan, utils whitesourceUtils) error {
```
@OliverNocon what do you think about handling MTAs here? This should be aligned with e.g. Fortify/Checkmarx.

stippi2 replied:
I am not quite sure what you mean.

OliverNocon replied:
Hope that I get the comment right.
The Unified Agent should in my opinion  be the toBe case for MTAs.
We should not fiddle around with the artifact components since this just creates a lot of issues and makes it hard to configure.

For now, current MTA handling is ok for me.
We should align with WhiteSource experts though to have only one scan per repo and thus for MTA have a proper config that the WhiteSource UA scans all MTA components with one call.

CCFenner commented on cmd/whitesourceExecuteScan.go:
```diff
@@ -184,65 +314,202 @@ func resolveProjectIdentifiers(config *ScanOptions, utils whitesourceUtils, sys
+
+// executeMavenScan constructs maven parameters from the given configuration, and executes the maven goal
+// "org.whitesource:whitesource-maven-plugin:19.5.1:update".
+func executeMavenScan(config *ScanOptions, scan *whitesourceScan, utils whitesourceUtils) error {
```
What about extracting this and related functions to an own package `pkg/whitesource/maven`?

CCFenner commented on cmd/whitesourceExecuteScan.go:
```diff
@@ -291,9 +574,125 @@ func writeWhitesourceConfigJSON(config *ScanOptions, utils whitesourceUtils, dev
 }
 
+// executeNpmScan iterates over all found npm modules and performs a scan in each one.
+func executeNpmScan(config *ScanOptions, scan *whitesourceScan, utils whitesourceUtils) error {
```
What about extracting this and related functions to an own package `pkg/whitesource/npm`?

OliverNocon commented on cmd/whitesourceExecuteScan.go:
```diff
@@ -184,65 +314,202 @@ func resolveProjectIdentifiers(config *ScanOptions, utils whitesourceUtils, sys
+	if err := autoGenerateWhitesourceConfig(config, utils); err != nil {
+		return err
+	}
+
```
besides downloading the agent we need to download also Java if not available in the runtime container.
Example can be found in the existing groovy step `whitesourceExecuteScan`

This is essential for scanning non Java projects with UA

stippi2 replied:
Yes, but I thought the decision was to not use the go-step for UA scans for the time being. The focus of this PR is to bypass the UA for MTA, Maven and NPM, not to fix the UA use-case.

daniel-kurzynski commented on cmd/whitesourceExecuteScan.go:
```diff
@@ -90,89 +126,183 @@ func newWhitesourceUtils() *whitesourceUtilsBundle {
+// and appends a new whitesource.Project with the given name.
+func (s *whitesourceScan) appendScannedProject(projectName string) error {
+	s.init()
+	projectName = projectName + " - " + s.projectVersion
```
I am confused. I would assume the projectVersion belongs to the project name. Why is projectName a parameter and projectVersion is already set before?

stippi2 replied:
Originally I had it like you would have expected (i.e. projectName parameter must already include the version), but then I thought it is bad style to duplicate the version appending code at the callers.

stippi2 replied:
I could rename the parameter to "moduleName" to make it more clear.

stippi2 replied:
Maybe I did address your question. All projects should be scanned with the same productVersion. There is code elsewhere in this step which also relies on this in order to find all projects belonging to the same scan as a sub-set of all projects in the same product. In order to enforce this, I thought it is a good idea to append the version only in one place. What I said about renaming the parameter to "moduleName" still applies.

stippi2 replied:
Renamed the parameter to "moduleName" to improve consistency of "what is what".

OliverNocon commented on cmd/whitesourceExecuteScan.go:
```diff
@@ -291,9 +574,125 @@ func writeWhitesourceConfigJSON(config *ScanOptions, utils whitesourceUtils, dev
+	utils.Stdout(ioutil.Discard)
+	defer utils.Stdout(log.Writer())
+
+	err := utils.RunExecutable("npm", "ls")
```
do we need this?
I think WhiteSource has an option to ignore `ls` errors but maybe I overlook s.th.

stippi2 replied:
You could be right, but I assumed the Cloud SDK implementation included this for a reason. I have not challenged whether this is valid or not, since I was not sure I could reproduce problems in my test projects.

daniel-kurzynski replied:
Unfortunately the npm ls ignore option did not work properly as far as I remember

daniel-kurzynski replied:
But maybe this has changed or was influenced by another problem.

daniel-kurzynski commented on cmd/whitesourceExecuteScan.go:
```diff
@@ -90,89 +126,183 @@ func newWhitesourceUtils() *whitesourceUtilsBundle {
 			config.ProductVersion = version
 		}
 	}
+	scan.projectVersion = config.ProductVersion
```
Project version is always equal to product version?

stippi2 replied:
Yes.

stippi2 replied:
Renamed it to scan.productVersion to make it consistent and clear.

OliverNocon reviewed (commented):
LGTM from what I can judge besides some comments

Open question: do we go ahead and address the comments in a follow-up PR or should it be done in this PR?

daniel-kurzynski commented on cmd/whitesourceExecuteScan.go:
```diff
@@ -184,65 +314,202 @@ func resolveProjectIdentifiers(config *ScanOptions, utils whitesourceUtils, sys
+// executeMTAScan executes a scan for the Java part with maven, and performs a scan for each NPM module.
+func executeMTAScan(config *ScanOptions, scan *whitesourceScan, utils whitesourceUtils) error {
+	log.Entry().Infof("Executing Whitesource scan for MTA project")
+	err := executeMavenScanForPomFile(config, scan, utils, "pom.xml")
```
I think you cannot alway assume there is a pom.xml file

stippi2 replied:
Ok, didn't know that before

daniel-kurzynski commented on cmd/whitesourceExecuteScan.go:
```diff
@@ -184,65 +314,202 @@ func resolveProjectIdentifiers(config *ScanOptions, utils whitesourceUtils, sys
+		return err
+	}
+
+	return executeNpmScan(config, scan, utils)
```
Same here. There might be MTA projects without npm ...

stippi2 replied:
Thanks!

OliverNocon reviewed (approved):
👍

daniel-kurzynski commented on cmd/whitesourceExecuteScan.go:
```diff
@@ -291,9 +574,125 @@ func writeWhitesourceConfigJSON(config *ScanOptions, utils whitesourceUtils, dev
+	}
+	exists, _ := utils.FileExists("package-lock.json")
+	if exists {
+		err = utils.FileRemove("package-lock.json")
```
I know we are currently doing it, but it is kind of a dirty hack which we should not do anymore ...

stippi2 replied:
I saw that the previous implementation for the npm scan passed a parameter `"ignoreNpmLsErrors":true` in the generated configuration. Our old implementation passed `"ignoreNpmLsErrors":false`, but implemented this check for "npm ls". Should we simply remove this code and always pass `"ignoreNpmLsErrors":true`? In our test-project, this check triggers for one of the npm-modules, so I know the code works. But I can test what happens if I flip the option.

daniel-kurzynski replied:
I would vote for "ignoreNpmLsErrors":true and remove this hack....

KevinHudemann reviewed (approved):
LGTM, although I can't really comment on functional correctness, thus additional approvals would be great!

stippi2 commented:
> LGTM from what I can judge besides some comments

Thanks!

> Open question: do we go ahead and address the comments in a follow-up PR or should it be done in this PR?

Daniel made some good points which would have been regressions, such as failing to scan MTAs that simply have no pom.xml part. I've addressed these and extended the tests.

Regarding the comments from Christopher on moving large chunks into the whitesource package, I have started that in a new branch, but would like to merge this PR now. My thinking is that the restructuring can be done step by step and it will be an improvement also in case I have to stop somewhere due to time constraints.

stippi2 commented:
I have had to restore the "npm ls" hack. I don't claim to understand fully what is going on. However, in my testing project, one of the npm modules triggers the code path where the "npm ls" check fails and "npm install" is performed after removing the node modules and package-lock.json. Without this, the npm whitesource plugin outputs "INFO: Ignore errors of 'npm ls'" which would be behavior according to the generated whitesource-config.json. However, the plugin execution then hangs after outputting "OK: Building dependencies report" to the log. I've repeated the test two times, each time it hung. It works with the hack restored. The npm plugin is no longer maintained, my thinking is that once the UA is ready, the dedicated npm scanning will be removed anyway, so I'd like to proceed with merging including the "npm ls" hack.

stippi2 commented:
/it
